main/src/ca/uwaterloo/flix/language/phase/TypeSimplifier.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2025 Jonathan Lindegaard Starup
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.language.ast.{Kind, Type}
+import ca.uwaterloo.flix.language.phase.typer.TypeConstraint
+import ca.uwaterloo.flix.language.phase.unification.EffUnification3
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+/** This type simplification is purely focused on user readability. */
+object TypeSimplifier {
+
+  /**
+    * Simplifies types and effects - intended for user readability.
+    *
+    * This function should never crash and works for any type, no matter how ill-kinded.
+    */
+  def simplify(tpe: Type): Type = {
+    // This might throw [[InternalCompilerException]] since it reuses unification infrastructure
+    // but since the error is so imprecise we cannot handle it here.
+    simplifyInternal(tpe)
+  }
+
+  /** Simplifies types inside `tc` with [[simplify]]. */
+  def simplifyConstraint(tc: TypeConstraint): TypeConstraint = tc match {
+    case TypeConstraint.Equality(tpe1, tpe2, prov) =>
+      TypeConstraint.Equality(simplify(tpe1), simplify(tpe2), prov)
+    case TypeConstraint.Trait(sym, tpe, loc) =>
+      TypeConstraint.Trait(sym, simplify(tpe), loc)
+    case TypeConstraint.Purification(sym, tpe1, tpe2, prov, nested) =>
+      TypeConstraint.Purification(sym, simplify(tpe1), simplify(tpe2), prov, nested.map(simplifyConstraint))
+    case TypeConstraint.Conflicted(tpe1, tpe2, prov) =>
+      TypeConstraint.Conflicted(simplify(tpe1), simplify(tpe2), prov)
+  }
+
+  /**
+    * Simplifies all effects in `tpe0`.
+    *
+    * Might throw [[InternalCompilerException]] for non-well-kinded effects because of [[EffUnification3.simplify]].
+    */
+  private def simplifyInternal(tpe0: Type): Type = tpe0 match {
+    case t if t.kind == Kind.Eff => EffUnification3.simplify(t)
```

Comment:
This converts to Zhegalkin form? That can be small, but its not simple or?

Should there not be some extra logic somewhere?

Reply:
I understand